LPIT Support on Zephyr

FelixWang47831 opened the pull request:
1. Provide counter driver based on lpit driver from NXP mcux-sdk-ng
2. Enable tests/drivers/counter/counter_nxp_lpit_api based on NXP mimxrt1180_evk, all channels test pass.

pdgendt commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,386 @@
+	return 0;
+}
+
+static DEVICE_API(counter, mcux_lpit_driver_api) = {
+		.start = mcux_lpit_start,
+		.stop = mcux_lpit_stop,
+		.get_value = mcux_lpit_get_value,
+		.set_top_value = mcux_lpit_set_top_value,
+		.get_pending_int = mcux_lpit_get_pending_int,
+		.get_top_value = mcux_lpit_get_top_value,
+		.get_freq = mcux_lpit_get_frequency,
+		.set_alarm = mcux_lpit_set_alarm,
+		.cancel_alarm = mcux_lpit_cancel_alarm
+};
```
```suggestion
static DEVICE_API(counter, mcux_lpit_driver_api) = {
	.start = mcux_lpit_start,
	.stop = mcux_lpit_stop,
	.get_value = mcux_lpit_get_value,
	.set_top_value = mcux_lpit_set_top_value,
	.get_pending_int = mcux_lpit_get_pending_int,
	.get_top_value = mcux_lpit_get_top_value,
	.get_freq = mcux_lpit_get_frequency,
	.set_alarm = mcux_lpit_set_alarm,
	.cancel_alarm = mcux_lpit_cancel_alarm,
};
```

pdgendt commented on tests/drivers/counter/counter_nxp_lpit_api/CMakeLists.txt:
Tests should be added in a separate commit.

pdgendt commented on dts/arm/nxp/nxp_rt118x.dtsi:
Changes to dts should be in a separate commit.

pdgendt commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,386 @@
+	.get_top_value = mcux_lpit_get_top_value,
+	.get_freq = mcux_lpit_get_frequency,
+	.set_alarm = mcux_lpit_set_alarm,
+	.cancel_alarm = mcux_lpit_cancel_alarm
```
```suggestion
	.cancel_alarm = mcux_lpit_cancel_alarm,
```

FelixWang47831 committed:
drivers: Counter: LPIT Support on Zephyr

1.Add dts bindings nxp,lpit-channel.yaml and nxp,lpit.yaml
2.Provide counter driver based on lpit driver from NXP mcux-sdk-ng

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
dts: arm: nxp: rt118x: add lpit instances for RT118X

Add lpit1, lpit2, lpit3 and all of the channels information

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
soc: nxp: imxrt: clock update for LPIT instances on RT118X

1. Configure clock source for lpit3 for imxrt118x devices
2. Support lpit in clock driver

Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

FelixWang47831 committed:
tests: drivers: counter_basic_api: Enable lpit test for mimxrt1180_evk

Enable test for all channels on both cm33 and cm7 core.
This test covers following test cases:
- test_set_top_value_with_alarm
- test_valid_function_without_alarm
- test_set_top_value_without_alarm
Signed-off-by: Felix Wang <fei.wang_3@nxp.com>

decsny reviewed (changes requested):
run the check compliance script, and explain why are you adding a test which is a copy paste of an existing test

pdgendt commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,382 @@
+			struct mcux_lpit_channel_data *data =
+				LPIT_CHANNEL_DATA(config->channels[channel_index]);
+
+			if (data->alarm_callback != NULL) {
```
No, this is a required code guideline, no implicit boolean conversions in `if` statements.

decsny replied:
ok

FelixWang47831 commented on soc/nxp/imxrt/imxrt118x/soc.c:
```diff
@@ -596,6 +596,15 @@ __weak void clock_init(void)
 #endif /* CONFIG_IMX_USDHC */
 
+#ifdef CONFIG_COUNTER_MCUX_LPIT
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(lpit3), okay)
+	/* Configure LPIT3 using SysPll3Div2 */
+	rootCfg.mux = kCLOCK_LPIT3_ClockRoot_MuxSysPll3Div2;
+	rootCfg.div = 3;
+	CLOCK_SetRootClock(kCLOCK_Root_Lpit3, &rootCfg);
+#endif
```
Because LPIT1 use BUS_AON, LPIT2 use BUS_WAKEUP, which have been configured before, has updated comment.

FelixWang47831 commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,382 @@
+	LPIT_StopTimer(config->base, channel_id);
+
+	return 0;
+}
```
Thanks for your advice. Updated.

manuargue commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,376 @@
+	 * you must disable the timer channel and then enable it again.
+	 */
+	mcux_lpit_stop(dev);
+	LPIT_SetTimerPeriod(config->base, channel, final_ticks);
```
same question as in this other pr: https://github.com/zephyrproject-rtos/zephyr/pull/93302#discussion_r2215774181
are you using the same counter register for both top callback and alarms? both APIs can be used simultaneously, so I don't think that's a good idea

FelixWang47831 replied:
Hi, yes, in short, LPIT does not have compare register,  it counts down from timer value register to 0, then trigger interrupt. 

In fact, this code implementation refers to counter_nxp_mrt.c and counter_nxp_pit.c. But they do not have  set_alarm and cancel_alarm function. I just wanted to ensure completeness and implement all the APIs, so I add them in LPIT driver.

I have carefully considered your comments, and based on how LPIT works, I believe  set_alarm and cancel_alarm are unnecessary. set_top_value and top callback are enough. What do you think?

manuargue replied:
> In fact, this code implementation refers to counter_nxp_mrt.c and counter_nxp_pit.c. But they do not have set_alarm and cancel_alarm function.

right, the nxp_pit have only top APIs implemented for the same reason I stated above.

> set_top_value and top callback are enough. What do you think?

according to your description of the timer, counts down from load value down to zero and trigger an interrupt, it seems more suitable for the top APIs to me than for the alarms.

FelixWang47831 commented on tests/drivers/counter/counter_nxp_lpit_api/src/main.c:
Considering suggestion from @manuargue, has removed  set_alarm and cancel_alarm API, now the tests/drivers/counter/counter_basic_api can cover LPIT API test. There is no need to provide this test, has been removed in the latest commit.

FelixWang47831 commented:
> run the check compliance script, and explain why are you adding a test which is a copy paste of an existing test

Hi @decsny , I have run check compliance script, and reuse existing test.

zejiang0jason commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,290 @@
+
+	if (running) {
+		/* Timer already enabled, check flags before resetting */
+		if (cfg->flags & COUNTER_TOP_CFG_DONT_RESET) {
```
should be `if ((cfg->flags & COUNTER_TOP_CFG_DONT_RESET) != 0)` ?

FelixWang47831 replied:
Fixed.

zejiang0jason commented on drivers/counter/counter_mcux_lpit.c:
```diff
@@ -0,0 +1,290 @@
+
+	for (int channel_index = 0; channel_index < config->num_channels; channel_index++) {
+
+		if (!(flags & (1U << channel_index))) {
```
should be `if ((flags & (1U << channel_index)) == 0)`?

zejiang0jason commented on drivers/counter/CMakeLists.txt:
```diff
@@ -33,6 +33,7 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_SNVS           counter_mcux_snv
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_TPM            counter_mcux_tpm.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_XEC                 counter_mchp_xec.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_LPTMR          counter_mcux_lptmr.c)
+zephyr_library_sources_ifdef(CONFIG_COUNTER_MCUX_LPIT          counter_mcux_lpit.c)
```
note the alignement

FelixWang47831 replied:
Fixed.
